src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp

```diff
@@ -107,26 +108,82 @@ CHIP_ERROR GeneralDiagosticsAttrAccess::Read(const ConcreteAttributePath & aPath
         return ReadNetworkInterfaces(aEncoder);
     }
     case RebootCount::Id: {
-        return ReadIfSupported(&PlatformManager::GetRebootCount, aEncoder);
+        return ReadIfSupported(&DeviceLayer::PlatformManager::GetRebootCount, aEncoder);
     }
     case UpTime::Id: {
-        return ReadIfSupported(&PlatformManager::GetUpTime, aEncoder);
+        return ReadIfSupported(&DeviceLayer::PlatformManager::GetUpTime, aEncoder);
     }
     case TotalOperationalHours::Id: {
-        return ReadIfSupported(&PlatformManager::GetTotalOperationalHours, aEncoder);
+        return ReadIfSupported(&DeviceLayer::PlatformManager::GetTotalOperationalHours, aEncoder);
     }
     case BootReasons::Id: {
-        return ReadIfSupported(&PlatformManager::GetBootReasons, aEncoder);
+        return ReadIfSupported(&DeviceLayer::PlatformManager::GetBootReasons, aEncoder);
     }
     default: {
         break;
     }
     }
     return CHIP_NO_ERROR;
 }
+
+class GeneralDiagnosticDelegate : public DeviceLayer::ConnectivityManagerDelegate, public DeviceLayer::PlatformManagerDelegate
+{
+
+    // Gets called when any network interface on the Node is updated.
+    void OnNetworkInfoChanged() override
+    {
+        ChipLogProgress(Zcl, "GeneralDiagnosticDelegate: OnNetworkInfoChanged");
+
+        for (uint16_t index = 0; index < emberAfEndpointCount(); index++)
+        {
+            if (emberAfEndpointIndexIsEnabled(index))
+            {
+                EndpointId endpointId = emberAfEndpointFromIndex(index);
+                if (endpointId == 0)
+                    continue;
+
+                if (emberAfContainsServer(endpointId, GeneralDiagnostics::Id))
+                {
+                    // If General Diagnostics cluster is implemented on this endpoint
+                    MatterReportingAttributeChangeCallback(endpointId, GeneralDiagnostics::Id,
+                                                           GeneralDiagnostics::Attributes::NetworkInterfaces::Id);
+                }
+            }
+        }
+    }
+
+    // Gets called when the device has been rebooted.
+    void OnDeviceRebooted() override
+    {
+        ChipLogProgress(Zcl, "GeneralDiagnosticDelegate: OnDeviceRebooted");
+
+        for (uint16_t index = 0; index < emberAfEndpointCount(); index++)
+        {
+            if (emberAfEndpointIndexIsEnabled(index))
+            {
+                EndpointId endpointId = emberAfEndpointFromIndex(index);
+
+                if (emberAfContainsServer(endpointId, GeneralDiagnostics::Id))
```

Comment:
This is a little weird in the sense that we go from index to endpoint id, but the first thing `emberAfContainsServer` ends up doing is getting the index from the endpoint id, which is all a huge waste of cycles.

We should have a function that takes an endpoint index and cluster id and returns whether that cluster is present on that endpoint....

Reply:
We need to get endpointID, it is needed by  MatterReportingAttributeChangeCallback

Reply:
That's fine.  Getting endpoint id from index is cheap.  But getting index from id is expensive, and that's what `emberAfContainsServer` ends up doing.

Hence my suggestion for a variant of `emberAfContainsServer` that takes an index instead of an endpoint id....